src/execution/index/art/art.cpp

```diff
@@ -204,6 +204,148 @@ void ART::GenerateKeys(DataChunk &input, vector<unique_ptr<Key>> &keys) {
+                idx_t depth) {
+
+	// test if we reached a leaf
+	auto prefix_start = depth;
+	while (keys[start_idx]->len != depth && keys[start_idx]->data[depth] == keys[end_idx]->data[depth]) {
+		depth++;
+	}
+
+	// we reached a leaf
+	if (depth == keys[start_idx]->len) {
+
+		// end_idx is inclusive
+		auto num_row_ids = end_idx - start_idx + 1;
+
+		// check for possible constraint violation
+		if ((IsUnique() || IsPrimary()) && num_row_ids != 1) {
+			throw ConstraintException("New data contains duplicates on indexed column(s)");
+		}
+
+		// new row ids of this leaf
+		auto new_row_ids = unique_ptr<row_t[]>(new row_t[num_row_ids]);
+		for (idx_t i = 0; i < num_row_ids; i++) {
+			new_row_ids[i] = row_ids[start_idx + i];
+		}
+
+		node = new Leaf(*keys[start_idx], prefix_start, move(new_row_ids), num_row_ids);
+
+	} else { // create a new node and recurse
+
+		// we will find at least two offset entries, otherwise we'd have reached a leaf
+		vector<BuildARTOffsets> offsets;
+		idx_t offset = start_idx;
+		for (idx_t i = start_idx + 1; i <= end_idx; i++) {
+			if (keys[i - 1]->data[depth] != keys[i]->data[depth]) {
+				BuildARTOffsets entry(offset, i - 1, keys[i - 1]->data[depth]);
+				offsets.push_back(entry);
+				offset = i;
+			}
+		}
+		BuildARTOffsets entry(offset, end_idx, keys[end_idx]->data[depth]);
+		offsets.push_back(entry);
+
+		if (offsets.size() < 5) {
+			node = new Node4();
+		} else if (offsets.size() < 17) {
+			node = new Node16();
+		} else if (offsets.size() < 49) {
+			node = new Node48();
+		} else {
+			node = new Node256();
+		}
+
+		auto prefix_length = depth - prefix_start;
+		node->prefix = Prefix(*keys[start_idx], prefix_start, prefix_length);
+
+		// recurse on each offset
+		for (idx_t i = 0; i < offsets.size(); i++) {
+			Node *child_node = nullptr;
+			Build(keys, row_ids, child_node, offsets[i].start, offsets[i].end, depth + 1);
+			Node::InsertChildNode(node, offsets[i].byte, child_node);
+		}
+	}
+}
+
+void ART::BuildAndMerge(IndexLock &lock, PayloadScanner &scanner, Allocator &allocator) {
+
+	auto payload_types = logical_types;
+	payload_types.emplace_back(LogicalType::ROW_TYPE);
+
+	auto skipped_all_nulls = false;
+	auto temp_art = make_unique<ART>(this->column_ids, this->unbound_expressions, this->constraint_type, this->db);
+	for (;;) {
+		DataChunk ordered_chunk;
+		ordered_chunk.Initialize(allocator, payload_types);
+		ordered_chunk.SetCardinality(0);
+		scanner.Scan(ordered_chunk);
+		if (ordered_chunk.size() == 0) {
+			break;
+		}
+
+		// get the key chunk and the row_ids vector
+		DataChunk chunk_row_ids;
+		ordered_chunk.Split(chunk_row_ids, ordered_chunk.ColumnCount() - 1);
+		auto &row_ids = chunk_row_ids.data[0];
+
+		D_ASSERT(row_ids.GetType().InternalType() == ROW_TYPE);
+		D_ASSERT(logical_types[0] == ordered_chunk.data[0].GetType());
+
+		// generate the keys for the given input
+		vector<unique_ptr<Key>> keys;
+		GenerateKeys(ordered_chunk, keys);
```

Comment:
The `GenerateKeys` now gets a datachunk, but internally uses a vector, we should change it to accept a vector, then you don't need all this splitting in there, you can directly use 
```
auto &row_ids = chunk.data[0];
auto &row_idskeys= chunk.data[1];
```

Reply:
`GenerateKeys` concatenates all key columns of a data chunk to a key, so I think that it should still take the whole key chunk as input.